### PR TITLE
[RUNTIME] Improve error messages for TypedPackedFunc

### PR DESCRIPTION
**Pull request by tkonolige**
This PR is an alternate to #7108, that captures the name of a TypedPackedFunc into a lambda instead of adding it as a field to the class. Right now, naming a TypedPackedFunc is optional, but I suggest we make it mandatory. I've already converted some of the places in the codebase to add a name (in this PR).

### CODE REVIEW

File: include/tvm/runtime/packed_func.h [hunk 1]
```diff
@@ -559,7 +569,18 @@ class TVMMovableArgValue_ : public TVMPODValue_ {
...
  private:
   /*! \return The arg value repr of the value. */
-  TVMArgValue AsArgValue() const { return TVMArgValue(value_, type_code_); }
+  TVMArgValue AsArgValue() const { return TVMArgValue(value_, type_code_, prefix_message_); }
+  /*! \brief Construct a message to provide context to any conversion errors in TVMPODValue_. */
+  static std::string error_string(const std::string& name, int argnum) {
+    if (name.size() == 0) {
+      return "";
+    }
+    std::string s = name + ": ";
```
Comment (areusch) [inline include/tvm/runtime/packed_func.h:578]: perhaps should add something here to indicate that `name` is a function?

`std::string s = std::string{"in calling function "} + s + ": ";`

File: include/tvm/ir/transform.h [hunk 2]
```diff
@@ -414,6 +414,65 @@ TVM_DLL Pass
 CreateModulePass(const runtime::TypedPackedFunc<IRModule(IRModule, PassContext)>& pass_func,
                  int opt_level, String name, Array<runtime::String> required);
 
+class ModulePassNode : public PassNode {
```
Comment (junrushao1994) [inline include/tvm/ir/transform.h:417]: I wonder why we need to move those pass definitions to a different place..Is this related to the error messages?

File: include/tvm/runtime/packed_func.h [hunk 3]
```diff
@@ -559,7 +569,18 @@ class TVMMovableArgValue_ : public TVMPODValue_ {
 
  private:
   /*! \return The arg value repr of the value. */
-  TVMArgValue AsArgValue() const { return TVMArgValue(value_, type_code_); }
+  TVMArgValue AsArgValue() const { return TVMArgValue(value_, type_code_, prefix_message_); }
+  /*! \brief Construct a message to provide context to any conversion errors in TVMPODValue_. */
+  static std::string error_string(const std::string& name, int argnum) {
```
Comment (junrushao1994) [inline include/tvm/runtime/packed_func.h:574]: Please do not write a static function in c++ header...Use inline to make sure only one copy of this function exists after linking.

File: include/tvm/ir/transform.h (hunk 2 above)
Comment (tkonolige) [reply include/tvm/ir/transform.h:417]: It is needed for `CreateModulePass` (https://github.com/apache/tvm/pull/7152/files#diff-99b9cb940bf19f00bb0d5956a156574f1f1940b026bff03fc0cb5f78c71d9faeR471). The new version of `CreateModulePass` needs to be templated so it accepts lambdas (and other functions) before they have been converted to `TypedPackedFunc`s. That way we can correctly set the name.

Comment (tqchen) [reply include/tvm/ir/transform.h:417]: I do think it is possible to keep things as they are(not moving to the header). Since the original rationale is to bring benefit of isolation. Ideally we don't want to cut one part to make things easier in another case. 

To resolve this, one way is to keep the API in their original form, while creating a separate function CreateModulePass_ that takes the `runtime::TypedPackedFunc` which constructs during the templated version. 



One possible way is to create an overloaded function of CreateModulePass_ that accepts the runtime::TypedPackedFunc which get calls into it.

Comment (tqchen) [reply include/tvm/ir/transform.h:417]: Additionally, this could be something that we can add back once the packed func land

File: include/tvm/runtime/packed_func.h [hunk 4]
```diff
@@ -442,13 +447,17 @@ class TVMPODValue_ {
...
-  TVMPODValue_(TVMValue value, int type_code) : value_(value), type_code_(type_code) {}
+  TVMPODValue_(TVMValue value, int type_code, const std::string& prefix_message = "")
+      : value_(value), type_code_(type_code), prefix_message_(prefix_message) {}
 
   /*! \brief The value */
   TVMValue value_;
   /*! \brief the type code */
   int type_code_;
+  /*! \brief Message to prefix any error message with */
+  std::string prefix_message_;
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h:460]: Adding String to each of the argument is a big overhead. We certainly don't want to construct such an error message for each function call even if no error is presented.

Instead, it might be good to think about alternatives. E.g. trying to be able to catch the errors in FFI functions and add name to the trace.

**Review by tqchen — CHANGES_REQUESTED** [review]

**Comment by tqchen** [comment]
Taking a closer look, i think we need to put more thoughts into it to make sure that the addition of name is optional, and we can get close to zero overhead in building these functions.

File: include/tvm/ir/transform.h (hunk 2 above)
Comment (tkonolige) [reply include/tvm/ir/transform.h:417]: We cannot take a `TypedPackedFunc` in `CreateModulePass` because the name needs to be provided to the constructor of `TypedPackedFunc`. We cannot set it after the fact.

File: include/tvm/runtime/packed_func.h (hunk 4 above)
Comment (tkonolige) [reply include/tvm/runtime/packed_func.h:460]: > Instead, it might be good to think about alternatives. E.g. trying to be able to catch the errors in FFI functions and add name to the trace.

This code gives error messages when the arguments cannot be converted correctly. It prints which argument does not match. I don't see how catching errors at the FFI can tell us which argument was incorrect.


The only other approach I can think of is manually converting each argument before we call the function. This would be a much larger change and would probably involve some template magic.

File: include/tvm/ir/transform.h (hunk 2 above)
Comment (tqchen) [reply include/tvm/ir/transform.h:417]: To clarify. 
- Have a function (not templated) `CreateModulePass_` that takes in `TypedPackedFunc`
- Create a templated function `CreateModulePass` that constructs the function(with name) and call into `CreateModulePass_`

File: include/tvm/runtime/packed_func.h (hunk 4 above)
Comment (tqchen) [reply include/tvm/runtime/packed_func.h:460]: I understand what this code does, want to point out the overhead in the current approach is too large that we might want to think about alternatives :)

The efficiency and clean part of the packedfunc mechanism really matters here that we might not want to include the feature because of the eager construction overhead.

I am not suggesting the FFI catching is the only way, but I do think more thoughts need to put into the process, for example, to have the error message report the i-th argument when things went wrong, while catching will show the name of the function.

Comment (tqchen) [reply include/tvm/runtime/packed_func.h:460]: Happy to chat more, i think there could be a better approach here

File: include/tvm/runtime/packed_func.h [hunk 5]
```diff
@@ -562,6 +570,37 @@ class TVMMovableArgValue_ : public TVMPODValue_ {
...
+/*!
+ * \brief Internal auxiliary struct for TypedPackedFunc to indicate a movable argument
+ *  with additional context information for better error reporting.
+ *
+ * \sa MovableArgValue_
+ * \note For internal development purpose only.
+ */
+class TVMMovableArgValueWithContext_ {
+ public:
+ public:
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h:582]: two public, document the arguments

File: include/tvm/runtime/packed_func.h [hunk 6]
```diff
@@ -562,6 +570,37 @@ class TVMMovableArgValue_ : public TVMPODValue_ {
...
+      LOG(FATAL) << "In function " << (name_ == nullptr ? "<anonymous>" : *name_)
+                 << ": error while converting argument " << arg_index_ << ": " << e.what();
+      throw "";
+    }
+  }
+
+ private:
+  TVMMovableArgValue_ value_;
+  int arg_index_;
+  const std::string* name_;
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h:601]: rename to opt_name_ indicate this is optional

File: include/tvm/runtime/packed_func.h [hunk 7]
```diff
@@ -562,6 +570,37 @@ class TVMMovableArgValue_ : public TVMPODValue_ {
...
+ */
+class TVMMovableArgValueWithContext_ {
+ public:
+ public:
+  TVMMovableArgValueWithContext_(TVMValue value, int type_code, int arg_index,
+                                 const std::string* name)
+      : value_(value, type_code), arg_index_(arg_index), name_(name) {}
+
+  template <typename T>
+  inline operator T() const {
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h:588]: when the function is inside a class body, iline is not needed

File: include/tvm/runtime/packed_func.h [hunk 8]
```diff
@@ -562,6 +570,37 @@ class TVMMovableArgValue_ : public TVMPODValue_ {
...
+    try {
+      return value_;
+    } catch (dmlc::Error& e) {
+      LOG(FATAL) << "In function " << (name_ == nullptr ? "<anonymous>" : *name_)
+                 << ": error while converting argument " << arg_index_ << ": " << e.what();
+      throw "";
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h:594]: add a comment about throw

File: include/tvm/runtime/packed_func.h [hunk 9]
```diff
@@ -562,6 +570,37 @@ class TVMMovableArgValue_ : public TVMPODValue_ {
   TVMArgValue AsArgValue() const { return TVMArgValue(value_, type_code_); }
 };
 
+/*!
+ * \brief Internal auxiliary struct for TypedPackedFunc to indicate a movable argument
+ *  with additional context information for better error reporting.
+ *
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h:576]: additional context information such as function name and argument index

File: include/tvm/runtime/packed_func.h [hunk 10]
```diff
@@ -526,7 +533,8 @@ class TVMArgValue : public TVMPODValue_ {
  */
 class TVMMovableArgValue_ : public TVMPODValue_ {
  public:
-  TVMMovableArgValue_(TVMValue value, int type_code) : TVMPODValue_(value, type_code) {}
+  TVMMovableArgValue_(TVMValue value, int type_code, const std::string& name = "", int argnum = -1)
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h:536]: this change is no longer needed

File: include/tvm/runtime/packed_func.h [hunk 11]
```diff
@@ -336,8 +337,11 @@ class TVMArgs {
 inline const char* ArgTypeCode2Str(int type_code);
 
 // macro to check type code.
+#define TVM_ICHECK_TYPE_CODE(CODE, T) \
+  ICHECK_EQ(CODE, T) << " expected " << ArgTypeCode2Str(T) << " but got " << ArgTypeCode2Str(CODE)
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h:341]: Given the macro is only used inside the packed_func, let us only keep one version. TVM_CHECK_TYPE_CODE should be fine for this case. (feel free to use ICHECK internally)

File: include/tvm/runtime/packed_func.h [hunk 12]
```diff
@@ -1213,21 +1252,22 @@ namespace detail {
...
+                                       name));
   }
 };
 
 template <typename R, int index, typename F>
 struct unpack_call_dispatcher<R, 0, index, F> {
   template <typename... Args>
-  TVM_ALWAYS_INLINE static void run(const F& f, const TVMArgs& args_pack, TVMRetValue* rv,
-                                    Args&&... unpacked_args) {
+  TVM_ALWAYS_INLINE static void run(const std::string* name, const F& f, const TVMArgs& args_pack,
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h:1269]: name -> optional_name

File: include/tvm/runtime/packed_func.h [hunk 13]
```diff
@@ -1297,14 +1339,18 @@ TypedPackedFunc<R(Args...)>::TypedPackedFunc(const TVMArgValue& value)
...
 
 template <typename R, typename... Args>
-TypedPackedFunc<R(Args...)>::TypedPackedFunc(TVMMovableArgValue_&& value)
+TypedPackedFunc<R(Args...)>::TypedPackedFunc(TVMMovableArgValueWithContext_&& value)
     : packed_(value.operator PackedFunc()) {}
 
 template <typename R, typename... Args>
 template <typename FType>
-inline void TypedPackedFunc<R(Args...)>::AssignTypedLambda(FType flambda) {
-  packed_ = PackedFunc([flambda](const TVMArgs& args, TVMRetValue* rv) {
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h:1306]: Let us keep an implementation when name is not provided, and no capturing is needed

File: include/tvm/runtime/packed_func.h [hunk 14]
```diff
@@ -1297,14 +1339,18 @@ TypedPackedFunc<R(Args...)>::TypedPackedFunc(const TVMArgValue& value)
...
-    detail::unpack_call<R, sizeof...(Args)>(flambda, args, rv);
+inline void TypedPackedFunc<R(Args...)>::AssignTypedLambda(FType flambda, std::string name) {
+  packed_ = PackedFunc([flambda, name](const TVMArgs& args, TVMRetValue* rv) {
+    if (args.size() != sizeof...(Args)) {
+      LOG(FATAL) << "Function " << (name.size() == 0 ? "<anonymous>" : name) << " expects "
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h:1350]: in this particular name is provided, and we can always print a name

**Review by tqchen — CHANGES_REQUESTED** [review]

File: include/tvm/ir/transform.h [hunk 15]
```diff
@@ -410,9 +410,30 @@ class Sequential : public Pass {
...
-CreateModulePass(const runtime::TypedPackedFunc<IRModule(IRModule, PassContext)>& pass_func,
-                 int opt_level, String name, Array<runtime::String> required);
+Pass CreateModulePass_(const runtime::TypedPackedFunc<IRModule(IRModule, PassContext)>& pass_func,
+                       int opt_level, String name, Array<runtime::String> required);
+
+/*
+ * \brief Create a module pass.
+ *
+ * \param pass_func The function that contains the optimization.
+ * \param opt_level The optimization level of the module pass.
```
Comment (tqchen) [inline include/tvm/ir/transform.h:420]: It would be great to leave these in a separate PR. Given the particular change will affect all PackedFunc calls. A smaller set of changes would be easier for others to review and take a look at

Comment (tkonolige) [reply include/tvm/ir/transform.h]: I've reverted these changes and will push them in a subsequent PR.

**Comment by tkonolige** [comment]
@tqchen @areusch @junrushao1994 I think this PR is in a good state to merge. Could you review again?

File: include/tvm/runtime/packed_func.h [hunk 16]
```diff
@@ -1377,7 +1449,7 @@ inline TObjectRef TVMPODValue_::AsObjectRef() const {
   using ContainerType = typename TObjectRef::ContainerType;
 
   if (type_code_ == kTVMNullptr) {
-    ICHECK(TObjectRef::_type_is_nullable)
+    CHECK(TObjectRef::_type_is_nullable)
```
Comment (areusch) [inline include/tvm/runtime/packed_func.h:1452]: shouldn't this be static_assert?

File: include/tvm/runtime/packed_func.h [hunk 17]
```diff
@@ -1240,16 +1293,19 @@ struct unpack_call_dispatcher<R, 0, index, F> {
...
   }
 };
 
 template <typename R, int nargs, typename F>
-TVM_ALWAYS_INLINE void unpack_call(const F& f, const TVMArgs& args, TVMRetValue* rv) {
-  ICHECK_EQ(nargs, args.size()) << "Expect " << nargs << " arguments but get " << args.size();
-  unpack_call_dispatcher<R, nargs, 0, F>::run(f, args, rv);
+TVM_ALWAYS_INLINE void unpack_call(const std::string* optional_name, const F& f,
+                                   const TVMArgs& args, TVMRetValue* rv) {
+  CHECK_EQ(nargs, args.size()) << (optional_name == nullptr ? "<anonymous>" : *optional_name)
```
Comment (areusch) [inline include/tvm/runtime/packed_func.h:1306]: maybe clarify what is the anonymous thing? i.e. `<anonymous function>` or like:
`CHECK_EQ(...) << "function " << (optional_name ...) << ": expected " << nargs << " arguments` ...

File: include/tvm/runtime/packed_func.h [hunk 18]
```diff
@@ -1297,14 +1353,30 @@ TypedPackedFunc<R(Args...)>::TypedPackedFunc(const TVMArgValue& value)
...
-TypedPackedFunc<R(Args...)>::TypedPackedFunc(TVMMovableArgValue_&& value)
+TypedPackedFunc<R(Args...)>::TypedPackedFunc(TVMMovableArgValueWithContext_&& value)
     : packed_(value.operator PackedFunc()) {}
 
+template <typename R, typename... Args>
+template <typename FType>
+inline void TypedPackedFunc<R(Args...)>::AssignTypedLambda(FType flambda, std::string name) {
+  packed_ = PackedFunc([flambda, name](const TVMArgs& args, TVMRetValue* rv) {
+    if (args.size() != sizeof...(Args)) {
+      LOG(FATAL) << "Function " << name << " expects " << sizeof...(Args) << " arguments, but "
```
Comment (areusch) [inline include/tvm/runtime/packed_func.h:1364]: maybe standardize the previous error message formatting with this one too

File: include/tvm/runtime/packed_func.h [hunk 19]
```diff
@@ -1297,14 +1353,30 @@ TypedPackedFunc<R(Args...)>::TypedPackedFunc(const TVMArgValue& value)
...
+  });
+}
+
 template <typename R, typename... Args>
 template <typename FType>
 inline void TypedPackedFunc<R(Args...)>::AssignTypedLambda(FType flambda) {
   packed_ = PackedFunc([flambda](const TVMArgs& args, TVMRetValue* rv) {
-    detail::unpack_call<R, sizeof...(Args)>(flambda, args, rv);
+    if (args.size() != sizeof...(Args)) {
+      LOG(FATAL) << "Function <anonymous> expects " << sizeof...(Args) << " arguments, but "
```
Comment (areusch) [inline include/tvm/runtime/packed_func.h:1376]: and here as well

**Review by areusch — APPROVED** [review]
a couple small things

File: include/tvm/runtime/packed_func.h (hunk 16 above)
Comment (tqchen) [reply include/tvm/runtime/packed_func.h:1452]: this should actually be a runtime check, note that constant folding will actually remove this completely

File: include/tvm/runtime/packed_func.h [hunk 20]
```diff
@@ -229,13 +229,13 @@ class TypedPackedFunc<R(Args...)> {
...
    * \tparam FLambda the type of the lambda function.
    */
-  template <typename FLambda, typename = typename std::enable_if<
-                                  std::is_convertible<FLambda,
-                                                      std::function<R(Args...)>>::value>::type>
-  TypedPackedFunc(const FLambda& typed_lambda) {  // NOLINT(*)
-    this->AssignTypedLambda(typed_lambda);
+  template <typename FLambda, typename = typename std::enable_if<std::is_convertible<
+                                  FLambda, std::function<R(Args...)>>::value>::type>
+  TypedPackedFunc(const FLambda& typed_lambda, std::string name = "<anonymous>") {  // NOLINT(*)
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h:237]: We need two constructors, one that has the string argument and another that does not.

The one that is anonymous should not capture any string content or construct a string

**Review by tqchen — CHANGES_REQUESTED** [review]

File: include/tvm/runtime/packed_func.h (hunk 16 above)
Comment (tkonolige) [reply include/tvm/runtime/packed_func.h:1452]: I think this is out of scope for this PR.

File: include/tvm/runtime/packed_func.h (hunk 17 above)
Comment (tkonolige) [reply include/tvm/runtime/packed_func.h]: Good point Andrew. I've changed the messages so they are all "Function <anonymous> expects....".

File: include/tvm/runtime/packed_func.h (hunk 20 above)
Comment (tkonolige) [reply include/tvm/runtime/packed_func.h]: Good catch. I totally forgot about that constructor.

**Review by tqchen — APPROVED** [review]

**Comment by tqchen** [comment]
@junrushao1994 please take another look and manage the PR. 
@tkonolige Thanks for keep improving the code through the reviewing process.

**Comment by junrushao1994** [comment]
Will do tonight!

File: include/tvm/runtime/packed_func.h [hunk 21]
```diff
@@ -562,6 +597,44 @@ class TVMMovableArgValue_ : public TVMPODValue_ {
...
+      : value_(value, type_code), arg_index_(arg_index), optional_name_(optional_name) {}
+
+  template <typename T>
+  operator T() const {
+    try {
+      return value_;  // implicit conversion happens here
+    } catch (dmlc::Error& e) {
+      LOG(FATAL) << "In function " << (optional_name_ == nullptr ? "<anonymous>" : *optional_name_)
+                 << ": error while converting argument " << arg_index_ << ": " << e.what();
+      throw "";  // never reached, LOG(FATAL) throws, but this silences a warning.
```
Comment (junrushao1994) [inline include/tvm/runtime/packed_func.h:628]: Looks like we don't have to throw anything haha

```suggestion
      throw;  // never reached, LOG(FATAL) throws, but this silences a warning.
```

**Review by junrushao1994 — COMMENTED** [review]
This PR overall looks good to me. Thank you so much for the effort making TVM errors more friendly :-)

BTW, probably out of the scope this this PR, given type mismatch is quite a common error across the codebase, I wonder if it makes sense to define a macro for this kind of error message, e.g.

```C++
#define TVM_ERROR_MSG_TYPE_MISMATCH(Expected, Actual)
   ("Expected type \"" + (Expected::_type_key) + ", but got type \"" + (Actual::_type_key) + '"')
```

**Comment by tkonolige** [comment]
Yeah, a type mismatch macro could be useful.

**Review by junrushao1994 — APPROVED** [review]
Looks good. Thanks Tristan!

**Comment by junrushao1994** [comment]
Thank you @tkonolige @tqchen @areusch! It is now merged!
